### DISCUSSION TITLE
June 2021 Release

### DISCUSSION THREAD

**Discussion by Gaming32** [Gaming32/ArrayV#57] [Announcements]
## New Sorts

+ Parallel Matrix Sort (idea by @Control55)
+ Rotate LSD Radix Sort (by @aphitorite)
+ Rotate MSD Radix Sort (by @aphitorite)
+ Simplistic Gravity Sort (By @McDude73)
+ Awkward Sort (by @aphitorite)
+ Complete Graph Sorting Network
+ Dandelion Sort (by yuji and @McDude73)
+ Fibonacci Gnome Sort (by @Yuri-chan2007)
+ Float Sort (by Lancewer and @McDude73)
+ Gambit Gnome Sort
+ Head Pull Sort (thanks to @PCBoyGames)
+ Linked Iterative Quick Sort, Left/Right Pointers (by @aphitorite)
+ Merry-Go-Round Sort (thanks to @PCBoyGames)
+ Odd-Even Weave Sort (High Prime) (thanks to @PCBoyGames)
+ Odd-Even Weave Sort (Low Prime) (thanks to @PCBoyGames)
+ Pattern-Defeating Gnome Sort (by @Yuri-chan2007)
+ Safe Bogo Sort (by @aphitorite)
+ Shove Sort (by frankblob and @PiotrGrochowski)
+ Split Center Sort (thanks to @PCBoyGames)
+ Swapless Bubble Sort (by @Gaming32)
+ Ternary Slow Sort (by @fungamer2-2)
+ X Pattern Sort (thanks to @PCBoyGames)
+ Circular Grail Sort (by @aphitorite)
+ Iterative Introspective Pseudo-Heap Sort (by @Yuri-chan2007)
+ Optimized Lazy Stable Sort
+ Pattern-Defeating Laziest Stable Sort (by @Yuri-chan2007)
+ Simple Hybrid Quick Sort (by @Yuri-chan2007)
+ Stackless Tim Sort (by @aphitorite)
+ Synchronous Sqrt Sort (by The Holy Grail Sort Project, implemented by @aphitorite)
+ Adaptive Insertion Sort
+ Binary Double Insertion Sort
+ Block Insertion Sort
+ Cocktail Shell Sort (Ciura Gaps) (by @fungamer2-2)
+ Pattern-Defeating Insertion Sort
+ Pancake Insertion Sort
+ Quick Sort (ternary, LL ptrs) (by @bingmann, thanks to @Gaming32)
+ Quick Sort (ternary, LR ptrs) (by @bingmann, thanks to @Gaming32)
+ Anarchy Sort (By Lancewer & @McDude73)
+ Binomial Heap Sort
+ Binomial Smooth Sort
+ Bottom-up Heap Sort (by @aphitorite)
+ Forced Stable Heap Sort (by @Yuri-chan2007)
+ Heavy Heap Sort
+ Min-Max Heap Sort
+ Out-of-Place Heap Sort (by @aphitorite)

## Changed Sorts

+ Set Parallel Bitonic Sort's unreasonable limit to 4096
+ Set Parallel Bose-Nelson Sorting Network's unreasonable limit to 4096
+ Set Parallel Odd-Even Merge Sort's unreasonable limit to 4096
+ Set Parallel Weave Sort's unreasonable limit to 4096
+ Remove indices array visualization from American Flag Sort
+ Remove indices array visualization from In-Place LSD Radix Sort
+ Remove indices array visualization from In-Place MSD Radix Sort
+ Fix comparisons in Smart Bogo Bogo Sort
+ Fix stability check and remove indices array visualization from Stackless American Flag Sort
+ Fix stability check for Stackless Binary Quick Sort
+ Fire Sort
+ Change highlights and delays for Gnome Sort
+ Change highlights and delays for Gnome Weave Sort (High Prime)
+ Change highlights and delays for Gnome Weave Sort (Low Prime)
+ Change delay for Quad Stooge Sort
+ Change delay for Slope Sort
+ Change delay for Stupid Fire Sort
+ Remove indices array visualization from Base-N Merge Sort
+ Redo Ecta Sort
+ Change median algorithm for Flan Sort
+ Rename Stable Buffered Merge Sort to Out-of-Place Buffered Merge Sort
+ Remi Sort
+ Rename Stable Third Merge
+ Fix Patience Sort comparisons
+ Shuffled Tree Sort
+ Set Parallel Rotate Merge Sort's unreasonable limit to 4096
+ Redo Strand Sort
+ Remove indices array visualization from Cube Root Quick Sort
+ Redo Asynchronous Sort

## Removed Sorts

+ Holy Grail Sort

## New Distributions

+ Modulo Function

## New Rotations

+ Bridge (by @Scandum, thanks to @aphitorite)

## New Features

+ Add sorting network visualizations (thanks to @Gaming32 and @aphitorite)

## Changed Features

+ Add seizure warning to Show Aux Arrays (as a temporary fix, until it no longer needs one) (thanks to @EmeraldBlock)
+ Re-add unsuccessful sort notifications to Stability Check
+ Change delay text to "Canceled" when canceled (used to say "0ms") (thanks to @EmeraldBlock)
+ Redo highlights system for *speed* (thanks to @EmeraldBlock)
+ Fix Cycle Reverse rotation's delay
+ Make sounds thread not make your CPU go brrrrr when idle
+ Change soundbank name to "Default (Yamaha XG Sound Set)" (was "Default (Roland SoundCanvas Sound Set)")

## Other changes

+ Added pull request template
+ Update workflow to use `actions/setup-java@v2.1.0` instead of `actions/setup-java@v1`
+ Make `.vscode/launch.json` match `run.bat` (thanks to @EmeraldBlock)

## Bug fixes

+ Fix final end sweep (in more cases)
+ Add visuals update on array re-initialize
+ Fix forgetting whether you had sound on
+ Fix rare errors when a sort contains a public inner class
+ Add Run All Quick Sorts
+ Fix "Pausedms" (thanks to @EmeraldBlock)

<hr /><em>This discussion was created from the release <a href='https://github.com/Gaming32/ArrayV-v4.0/releases/tag/v21.5.0'>June 2021</a>.</em>

**Comment by EmeraldBlock** [comment]
Pausedms

**Reply by Gaming32** [reply]
Thanks for fixing that by the way 😅
